File: java/src/main/MakeSumDivisibleByP.java
import util.ArrayUtils;

import java.util.HashMap;
import java.util.Map;

import static util.ArrayUtils.array;
import static util.ResourceUtils.loadTestcase;

/**
 * 1590. 使数组和能被 P 整除：https://leetcode.cn/problems/make-sum-divisible-by-p
 */
public class MakeSumDivisibleByP {

    public int minSubarray(int[] nums, int p) {
        long t = 0;
        for (int num : nums) {
            t += num;
        }
        if (t < p) {
            return -1;
        }
        int k = (int) (t % p);
        if (k == 0) {
            return 0;
        }
        int n = nums.length;
        Map<Integer, Integer> last = new HashMap<>(n, 1);
        int s = 0;
        int ans = Integer.MAX_VALUE;
        last.put(s, -1);
        for (int i = 0; i < n; i++) {
            s = (s + nums[i]) % p;
            last.put(s, i);
            ans = Math.min(ans, i - last.getOrDefault((s - k + p) % p, -n));
        }
        return ans < n ? ans : -1;
    }

    public static void main(String[] args) {
        MakeSumDivisibleByP app = new MakeSumDivisibleByP();
        System.out.println(app.minSubarray(ArrayUtils.array("[6,3,5,2]"), 9)); // 2
        System.out.println(app.minSubarray(ArrayUtils.array("[4,4,2]"), 7)); // -1
        System.out.println(app.minSubarray(array(loadTestcase("1590-testcase-1.txt")), 949467102));
        System.out.println(app.minSubarray(array("[8,32,31,18,34,20,21,13,1,27,23,22,11,15,30,4,2]"), 148));
    }

}

File: c/make-sum-divisible-by-p.c
/**
 * 1590. 使数组和能被 P 整除：https://leetcode.cn/problems/make-sum-divisible-by-p
 */

#include <stdio.h>
#include <stdlib.h>
#include "uthash.h"

#define min(a,b) (((a) < (b)) ? (a) : (b))

typedef struct {
    int key;
    int value;
    UT_hash_handle hh;
} MapItem;

static void put(MapItem **map, int key, int value) {
    MapItem *item;
    HASH_FIND_INT(*map, &key, item);
    if (item == NULL) {
        item = malloc(sizeof(MapItem));
        item->key = key;
        item->value = value;
        HASH_ADD_INT(*map, key, item);
    } else {
        item->value = value;
    }
}

static MapItem* get(MapItem **map, int key) {
    MapItem *item;
    HASH_FIND_INT(*map, &key, item);
    return item;
}

static void remove_key(MapItem **map, int key) {
    MapItem *item;
    HASH_FIND_INT(*map, &key, item);
    if (item) {
        HASH_DEL(*map, item);
        free(item);
    }
}

static int size(const MapItem **map) {
    return HASH_COUNT(*map);
}

static void free_all(MapItem **map) {
    MapItem *curr, *tmp;
    HASH_ITER(hh, *map, curr, tmp) {
        HASH_DEL(*map, curr);
        free(curr);
    }
}


int minSubarray(int* nums, int numsSize, int p) {
    long t = 0L;
    for (int i = 0; i < numsSize; i++) {
        t += nums[i];
    }
    if (t < p) {
        return -1;
    }
    const int k = t % p;
    if (k == 0) {
        return 0;
    }

    MapItem *last = NULL;
    int ans = numsSize;
    int s = 0;
    put(&last, s, -1);
    for (int i = 0; i < numsSize; i++) {
        s = (s + nums[i]) % p;
        put(&last, s, i);
        const MapItem * item = get(&last, (s - k + p) % p);
        const int j = item ? item->value : -numsSize;
        ans = min(ans, i - j);
    }
    free_all(&last);
    return ans < numsSize ? ans : -1;
}
